### PR TITLE
update connector to support custom URI's

### PR DESCRIPTION
**Pull request by shorsher**
in a chain with https://github.com/hyperledger/firefly-tokens-erc20-erc721/pull/60

### CODE REVIEW

File: src/tokens/tokens.interfaces.ts [hunk 1]
```diff
@@ -291,6 +295,10 @@ export class TokenTransfer {
   @IsOptional()
   amount?: string;
 
+  @ApiProperty()
+  @IsOptional()
+  uri?: string;
```
Comment (awrichar) [inline src/tokens/tokens.interfaces.ts:300]: I think this should only be settable on "mint", not transfer/burn - right?

File: src/tokens/tokens.service.ts [hunk 2]
```diff
@@ -432,7 +444,8 @@ export class TokensService {
   }
 
   async supportsData(address: string, type: TokenType) {
-    const iid = type === TokenType.FUNGIBLE ? ERC20WithDataIID : ERC721WithDataIID;
+    const withDataIID = await this.supportsNFTUri(address, false) ? ERC721WithDataUriIID : ERC20WithDataIID
+    const iid = type === TokenType.FUNGIBLE ? ERC20WithDataIID : withDataIID;
```
Comment (awrichar) [inline src/tokens/tokens.service.ts:448]: This logic might be correct but seems a little hard to follow... maybe a `switch` would be clearer?

File: src/tokens/tokens.service.ts [hunk 3]
```diff
@@ -199,7 +209,9 @@ export class TokensService {
     public http: HttpService,
     private eventstream: EventStreamService,
     private proxy: EventStreamProxyGateway,
-  ) {}
+  ) {
+    this.uriCache = new LRUCache({ max: 500 });
```
Comment (awrichar) [inline src/tokens/tokens.service.ts:213]: nit: this isn't really a cache of URIs, but a cache of whether a given contract has URI support - right?

File: src/tokens/tokens.service.ts (hunk 2 above)
Comment (shorsher) [reply src/tokens/tokens.service.ts]: tried to make it clearer with a `switch` and better naming

File: src/tokens/tokens.service.ts [hunk 4]
```diff
@@ -445,6 +469,25 @@ export class TokensService {
...
 
+  async supportsNFTUri(address: string, factory: boolean) {
+    const support = this.uriSupportCache.get(address);
+    if (support) {
+      return support;
+    }
+
+    try {
+      const result = await this.query(address, supportsInterfaceABI, factory ? [TokenFactoryIID] :[ERC721WithDataUriIID]);
+      this.logger.log(`Querying extra data support on contract '${address}': ${result.output}`);
```
Comment (awrichar) [inline src/tokens/tokens.service.ts:480]: Should these logs say "URI support"?

File: src/tokens/tokens.service.ts [hunk 5]
```diff
@@ -432,7 +444,19 @@ export class TokensService {
   }
 
   async supportsData(address: string, type: TokenType) {
-    const iid = type === TokenType.FUNGIBLE ? ERC20WithDataIID : ERC721WithDataIID;
+    const nftIID = await this.supportsNFTUri(address, false) ? ERC721WithDataUriIID : ERC20WithDataIID
```
Comment (awrichar) [inline src/tokens/tokens.service.ts:447]: If `supportsNFTURI` returns true, doesn't that mean it also supports data? Seems like we can short-circuit and return much earlier.

File: src/tokens/tokens.service.ts [hunk 6]
```diff
@@ -532,13 +575,20 @@ export class TokensService {
     if (method === undefined) {
       throw new BadRequestException('Failed to parse factory contract ABI');
     }
+    const params = [dto.name, dto.symbol, isFungible, encodedData];
+    const uri = await this.supportsNFTUri(this.factoryAddress, true)
+    if (uri) {
+      // supply empty string is URI isn't provided
```
Comment (awrichar) [inline src/tokens/tokens.service.ts:581]: `is` -> `if`

File: test/suites/erc721.ts [hunk 7]
```diff
@@ -107,13 +115,48 @@ export default (context: TestContext) => {
...
+        data: `{"tx":${TX}}`,
+        poolLocator: `address=${CONTRACT_ADDRESS}&schema=${ERC721_WITH_DATA_SCHEMA}&type=${TokenType.NONFUNGIBLE}`,
+        standard: 'ERC721',
+        type: TokenType.NONFUNGIBLE,
+        symbol: SYMBOL,
+        info: {
+          name: NAME,
+          address: CONTRACT_ADDRESS,
+          schema: ERC721_WITH_DATA_SCHEMA,
+        },
```
Comment (awrichar) [inline test/suites/erc721.ts:148]: Should we consider adding the base URI to the `info` returned here?

**Review by awrichar — COMMENTED** [review]
Left a few comments for consideration. Might want to double-check `npm run lint` and `npm run format` too.
